src/estimators/stratified_sketch.py

```diff
@@ -0,0 +1,242 @@
+              that.sketches[str(k + 1)], self.sketch_intersection(
+                  this.sketches[ONE_PLUS], that.sketches[str(k + 1)])))
+
+      # Calculate A(i) & B(k-i)
+      for i in range(k):
+        merged = self.sketch_union(
+            merged, self.sketch_intersection(
+                this.sketches[str(i + 1)], that.sketches[str(k - i)]))
+      merged_sketch.sketches[str(k + 1)] = merged
+
+    # Calculate Merged(max_freq)
+    merged = this.sketches[str(max_freq)]
+    rest = that.sketches[ONE_PLUS]
+    for k in range(max_freq - 1):
+      merged = self.sketch_union(
+          merged,
+          self.sketch_intersection(this.sketches[str(max_freq - k - 1)],
+                                   rest))
+      rest = self.sketch_difference(rest, that.sketches[str(k + 1)])
+    merged = self.sketch_union(
+        merged,
+        self.sketch_difference(
+            that.sketches[str(max_freq)],
+            self.sketch_intersection(
+                that.sketches[str(max_freq)],
+                this.sketches[ONE_PLUS])))
+    merged_sketch.sketches[str(max_freq)] = merged
+
+    # Calculate Merged(1+)
+    merged_one_plus = None
+    for k in range(max_freq):
+      merged_one_plus = self.sketch_union(merged_one_plus,
+                                          merged_sketch.sketches[str(k + 1)])
+    merged_sketch.sketches[ONE_PLUS] = merged_one_plus
+
+    return merged_sketch
+
+class SequentialEstimator(object):
+  """Sequential frequency estimator."""
+
+  def __init__(self, sketch_operator):
+    self.pairwise_estimator = PairwiseEstimator(sketch_operator)
+
+  def __call__(self, sketches_list):
+    return self.merge_sketches(sketches_list)
+
+  def merge_sketches(self, sketches_list):
+    return functools.reduce(self.pairwise_estimator.merge_sketches,
+                            sketches_list)
+
+class ExactMultiSetOperation(object):
+
+  @classmethod
+  def union(cls, this, that):
+    if this is None:
+      return copy.deepcopy(that)
+    if that is None:
+      return copy.deepcopy(this)
+    result = copy.deepcopy(this)
+    result_key_set = set(result.ids().keys())
+    that_key_set = set(that.ids().keys())
+    result._ids = {x:1 for x in result_key_set.union(that_key_set)}
+    return result
+
+  @classmethod
+  def intersection(cls, this, that):
+    if this is None or that is None:
+      return None
+    result = copy.deepcopy(this)
+    result_key_set = set(result.ids().keys())
+    that_key_set = set(that.ids().keys())
+    result._ids = {x:1 for x in result_key_set.intersection(that_key_set)}
+    return result
+
+  @classmethod
+  def difference(cls, this, that):
+    if this is None:
+      return None
+    if that is None:
+      return copy.deepcopy(this)
+    result = copy.deepcopy(this)
+    result_key_set = set(result.ids().keys())
+    that_key_set = set(that.ids().keys())
+    result._ids = {x:1 for x in result_key_set.difference(that_key_set)}
+    return result
```

Comment:
Why not add these methods to ExactMultiSet?  This would reduce the number of classes by one, and it would keep all of the code relating to ExactMultiSet in one place.

Reply:
@huangxichen1 we had this conversation and I think your reasoning is to keep it this way because " The union and set_difference are applied to a specific sets relationship, which can be different from a more generic situations"
 
Matthew, maybe we should keep it this way because the other sketches will need other operators which  @huangxichen1 is going to work on.